PROYECTO 1


El objetivo del proyecto es comparar el desempeño de diferentes modelos de aprendizaje
automático en un problema de regresión. A saber, la predicción del precio de casas en una
región de Estados Unidos, con base en 12 atributos.

1. Cargar el dataset HousingData, el cual se proveerá junto con un diccionario de los atributos. La variable objetivo es MEDV.

In [133]:
#importar paquetes
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression #Algoritmo de regresión lineal
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import Ridge
from sklearn.linear_model import LassoCV
from sklearn.linear_model import Lasso

In [134]:
#TODO Leer el conjunto de datos y darle el nombre df
df=pd.read_csv("HousingData1.csv")
df

Unnamed: 0     CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  \
0             0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1   
1             1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2   
2             2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2   
3             3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3   
4             4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3   
..          ...      ...   ...    ...   ...    ...    ...   ...     ...  ...   
501         501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1   
502         502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1   
503         503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1   
504         504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1   
505         505  0.04741   0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1   

     TAX  PTRATIO  LSTAT  MEDV  
0    296     15.3   4.98  24.0  
1    242     17.8   9.14  21.6  
2    242     17.8   4.03  34.7  
3    222     18.7   2.94  33.4  
4    222     18.7    NaN  36.2  
..   ...      ...    ...   ...  
501  273     21.0    NaN  22.4  
502  273     21.0   9.08  20.6  
503  273     21.0   5.64  23.9  
504  273     21.0   6.48  22.0  
505  273     21.0   7.88  11.9  

[506 rows x 14 columns]

2. Examinar y realizar una breve descripción de los tipos de variable o cualquier observación que considere pertinente. Remueva filas con valores ausentes si en algún punto lo considera necesario.

In [135]:
#TODO5 Remover filas con valores ausentes.
df=df.dropna()
df

Unnamed: 0     CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  \
0             0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1   
1             1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2   
2             2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2   
3             3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3   
5             5  0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622    3   
..          ...      ...   ...    ...   ...    ...    ...   ...     ...  ...   
499         499  0.17783   0.0   9.69   0.0  0.585  5.569  73.5  2.3999    6   
500         500  0.22438   0.0   9.69   0.0  0.585  6.027  79.7  2.4982    6   
502         502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1   
503         503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1   
504         504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1   

     TAX  PTRATIO  LSTAT  MEDV  
0    296     15.3   4.98  24.0  
1    242     17.8   9.14  21.6  
2    242     17.8   4.03  34.7  
3    222     18.7   2.94  33.4  
5    222     18.7   5.21  28.7  
..   ...      ...    ...   ...  
499  391     19.2  15.10  17.5  
500  391     19.2  14.33  16.8  
502  273     21.0   9.08  20.6  
503  273     21.0   5.64  23.9  
504  273     21.0   6.48  22.0  

[394 rows x 14 columns]

In [136]:
df=df.drop(df.columns[[0]], axis=1)
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
5    0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
499  0.17783   0.0   9.69   0.0  0.585  5.569  73.5  2.3999    6  391   
500  0.22438   0.0   9.69   0.0  0.585  6.027  79.7  2.4982    6  391   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   

     PTRATIO  LSTAT  MEDV  
0       15.3   4.98  24.0  
1       17.8   9.14  21.6  
2       17.8   4.03  34.7  
3       18.7   2.94  33.4  
5       18.7   5.21  28.7  
..       ...    ...   ...  
499     19.2  15.10  17.5  
500     19.2  14.33  16.8  
502     21.0   9.08  20.6  
503     21.0   5.64  23.9  
504     21.0   6.48  22.0  

[394 rows x 13 columns]

In [137]:
#Explorar el dataset
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
5  0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622    3  222     18.7   

   LSTAT  MEDV  
0   4.98  24.0  
1   9.14  21.6  
2   4.03  34.7  
3   2.94  33.4  
5   5.21  28.7

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 394 entries, 0 to 504
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     394 non-null    float64
 1   ZN       394 non-null    float64
 2   INDUS    394 non-null    float64
 3   CHAS     394 non-null    float64
 4   NOX      394 non-null    float64
 5   RM       394 non-null    float64
 6   AGE      394 non-null    float64
 7   DIS      394 non-null    float64
 8   RAD      394 non-null    int64  
 9   TAX      394 non-null    int64  
 10  PTRATIO  394 non-null    float64
 11  LSTAT    394 non-null    float64
 12  MEDV     394 non-null    float64
dtypes: float64(11), int64(2)
memory usage: 43.1 KB


En este paso podemos observar el tipo que es cada variable.

In [139]:
df.value_counts()

CRIM      ZN    INDUS  CHAS  NOX    RM     AGE    DIS     RAD  TAX  PTRATIO  LSTAT  MEDV
0.00632   18.0  2.31   0.0   0.538  6.575  65.2   4.0900  1    296  15.3     4.98   24.0    1
1.34284   0.0   19.58  0.0   0.605  6.066  100.0  1.7573  5    403  14.7     6.43   24.3    1
1.25179   0.0   8.14   0.0   0.538  5.570  98.1   3.7979  4    307  21.0     21.02  13.6    1
1.23247   0.0   8.14   0.0   0.538  6.142  91.7   3.9769  4    307  21.0     18.72  15.2    1
1.20742   0.0   19.58  0.0   0.605  5.875  94.6   2.4259  5    403  14.7     14.43  17.4    1
                                                                                           ..
0.11460   20.0  6.96   0.0   0.464  6.538  58.7   3.9175  3    223  18.6     7.73   24.4    1
0.11432   0.0   8.56   0.0   0.520  6.781  71.3   2.8561  5    384  20.9     7.67   26.5    1
0.11132   0.0   27.74  0.0   0.609  5.983  83.5   2.1099  4    711  20.1     13.35  20.1    1
0.11069   0.0   13.89  1.0   0.550  5.951  93.8   2.8893  5    27

3. Separe el conjunto de datos en atributos X y variable objetivo y. Luego separe éstos en conjuntos de entrenamiento y testeo, con una proporción de 75-25.


In [140]:
##Separar la columna objetivo.
X=df.drop('MEDV',axis=1)
y=df['MEDV']

In [141]:
#TODO8 Remover columnas irrelevantes
X=X.drop(['CHAS'],axis=1)
X

CRIM    ZN  INDUS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0    0.00632  18.0   2.31  0.538  6.575  65.2  4.0900    1  296     15.3   
1    0.02731   0.0   7.07  0.469  6.421  78.9  4.9671    2  242     17.8   
2    0.02729   0.0   7.07  0.469  7.185  61.1  4.9671    2  242     17.8   
3    0.03237   0.0   2.18  0.458  6.998  45.8  6.0622    3  222     18.7   
5    0.02985   0.0   2.18  0.458  6.430  58.7  6.0622    3  222     18.7   
..       ...   ...    ...    ...    ...   ...     ...  ...  ...      ...   
499  0.17783   0.0   9.69  0.585  5.569  73.5  2.3999    6  391     19.2   
500  0.22438   0.0   9.69  0.585  6.027  79.7  2.4982    6  391     19.2   
502  0.04527   0.0  11.93  0.573  6.120  76.7  2.2875    1  273     21.0   
503  0.06076   0.0  11.93  0.573  6.976  91.0  2.1675    1  273     21.0   
504  0.10959   0.0  11.93  0.573  6.794  89.3  2.3889    1  273     21.0   

     LSTAT  
0     4.98  
1     9.14  
2     4.03  
3     2.94  
5     5.21  
..     ...  
499  15.10  
500  14.33  
502   9.08  
503   5.64  
504   6.48  

[394 rows x 11 columns]

In [142]:
#Separar en train y test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0, test_size=0.25, train_size=0.75, shuffle=True)

4. Entrene un modelo de regresión lineal usando el conjunto de entrenamiento. Determine su coeficiente de determinación (R2 o, en el caso de SKLearn el score). Extraiga los coeficientes o parámetros predichos por el modelo usando el atributo coef_

In [143]:
# Modelo de regresión lineal de SciKit Learn
#POR MIRAR CON EL Q SE ENTRENA Y ESAS COSAS
regr = LinearRegression()
regr.fit(X_train, y_train)
print("score = ", regr.score(X_test, y_test))

score =  0.6467476303419486


In [144]:
print('slope:', regr.coef_) 
print('intercept:', regr.intercept_)

slope: [-1.00786413e-01  4.44912105e-02  7.89424366e-02 -2.22991830e+01
  4.72052256e+00 -1.72531916e-02 -1.55743542e+00  2.60567073e-01
 -1.49306629e-02 -9.74971101e-01 -3.87625432e-01]
intercept: 37.77489522513062


5. Mencione cuáles atributos obtuvieron un coeficiente con mayor magnitud. Éstos son los que más afectan a la variable objetivo.

Los atributos que obtuvieron un coeficiente con mayor magnitud son: ZN(4.449), INDUS(7.89), RM(4.72), PTRATIO(-9.74) y LSTAT(-3.87).

6. Usando la validación cruzada de Ridge (RidgeCV) busque un valor óptimo para el parámetro α, de forma similar al taller1.


In [145]:
#Se Halla el alfa mas optimo

alfa = RidgeCV(alphas=[1e-3, 1e-2, 1e-1, 1]).fit(X_train, y_train)
alfa.alpha_

0.01

7. Realice una regresión de Ridge usando el parámetro α hallado en el punto anterior.

In [146]:
#Modelo de regresión de Ridge usando el alfa hallado anteriormente
model= Ridge(alfa.alpha_, fit_intercept=False)
model.fit(X_train,y_train)
print("score = ", model.score(X_test, y_test))

score =  0.5985523762298866


8. Repita los dos puntos anteriores para regresión de Lasso.

In [150]:
#Se Halla el alfa mas optimo para Lasso

alfa_lasso = LassoCV(alphas=[1e-3, 1e-2, 1e-1, 1]).fit(X_train, y_train)
alfa_lasso.alpha_

0.001

In [151]:
#Modelo de regresión de lASSO usando el alfa hallado anteriormente
model_Lasso= Lasso(alfa_lasso.alpha_, fit_intercept=False)
model_Lasso.fit(X_train,y_train)
print("score = ", model_Lasso.score(X_test, y_test))

score =  0.5990240216779897


9. ¿Aplicar regresión de Ridge y Lasso mejoró el score de ésta regresion? ¿Por qué cree que ocurrió de esta manera?




referencias
Ridge= https://machinelearningmastery.com/ridge-regression-with-python/


Ridgecv= https://www.cienciadedatos.net/documentos/py14-ridge-lasso-elastic-net-python.html

https://www.cienciadedatos.net/documentos/py06_machine_learning_python_scikitlearn.html


https://bookdown.org/content/2274/metodos-de-clasificacion.html

10. Aplique árboles de decisión de regresión para predecir la variable objetivo con distintos
valores de los par´ametros max leaf nodes y max depth. Qu´edese con el modelo que d´e
un mayor score.